<div class="clearfix" style="padding: 10px; padding-left: 0px">
<a href="http://bombora.com"><img src="https://app.box.com/shared/static/e0j9v1xjmubit0inthhgv3llwnoansjp.png" width="200px" class="pull-right" style="display: inline-block; margin: 5px; vertical-align: middle;"></a>
<h1> Bombora Data Science: <br> *Interview Exam* </h1>
</div>

<img width="200px" src="https://app.box.com/shared/static/15slg1mvjd1zldbg3xkj9picjkmhzpa5.png">

---
# Welcome

Welcome! This notebook contains interview exam questions referenced in the *Instructions* section in the `README.md`—please read that first, *before* attempting to answer questions here.

<div class="alert alert-info" role="alert" style="margin: 10px">
<p style="font-weight:bold">ADVICE</p>
<p>*Do not* read these questions, and panic, *before* reading the instructions in `README.md`.</p>
</div>

<div class="alert alert-warning" role="alert" style="margin: 10px">
<p style="font-weight:bold">WARNING</p>

<p>If using <a href="https://try.jupyter.org">try.jupyter.org</a> do not rely on the server for anything you want to last - your server will be <span style="font-weight:bold">deleted after 10 minutes of inactivity</span>. Save often and rember download notebook when you step away (you can always re-upload and start again)!</p>
</div>


## Have fun!

Regardless of outcome, getting to know you is important. Give it your best shot and we'll look forward to following up!

# Exam Questions

## 1. Algo + Data Structures

### Q 1.1: Fibionacci
![fib image](https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Fibonacci_spiral_34.svg/200px-Fibonacci_spiral_34.svg.png)

#### Q 1.1.1
Given $n$ where $n \in \mathbb{N}$ (i.e., $n$ is an integer and $n > 0$), write a function `fibonacci(n)` that computes the Fibonacci number $F_n$, where $F_n$ is defined by the recurrence relation:

$$ F_n = F_{n-1} + F_{n-2}$$

with initial conditions of:

$$ F_1 = 1,  F_2 = 1$$

In [23]:
def fibonacci(n):
    if (n <= 1):
        return n
    if (n == 2):
        return 1

    cur = 0
    f_1 = 1
    f_2 = 1

    # iteratively compute the Fibonacci number
    for i in range(2, n):
        cur = f_1 + f_2
        f_2 = f_1
        f_1 = cur
    return cur

assert fibonacci(1) == 1
assert fibonacci(3) == 2
assert fibonacci(10) == 55

#### Q 1.1.2
What's the complexity of your implementation?

Time complexity $O(n)$: The `for loop` iterates $n-2$ times, which is proportionate to $n$, the Fibonacci number we are trying to compute.

Space complexity $O(1)$: No data structures were used and the unit of space each variable occupies does not change.

#### Q 1.1.3
Consider an alternative implementation to compute Fibonacci number $F_n$ and write a new function, `fibonacci2(n)`.

In [24]:
def fibonacci(n):
    # base case
    if n <= 1:
        return n
    
    return fibonacci(n-1) + fibonacci(n-2)

assert fibonacci(1) == 1
assert fibonacci(3) == 2
assert fibonacci(10) == 55

#### Q 1.1.4
What's the complexity of your implementation?

Time complexity $O(2^n)$: Each recursion call initiates 2 recursion calls with depth equal to $n$, thus total number of calls is $2^0+2^1+2^2+...+2^{n-1} = 2^{n} - 1$, which is proportionate to $2^n$.

Space complexity $O(n)$: Since `fibonacci(n-1)` gets evaluated completely before `fibonacci(n-2)`, there will never be more than n stack calls at any given time.

#### Q 1.1.5
What are some examples of optimizations that could improve computational performance?


We can use memoization to store computed answers into a dictionary, and look up compute values at each recursion call. With memoizatoin, the function does not repeat the work to calculate known values and time complexity is reduced to $O(n)$ since the function only needs to compute each fibonacci number once and look up of dictionay is $O(1)$. Space complexity stays the same since $O(n+n) = O(n)$.

### Q 1.2: Linked List
![ll img](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6d/Singly-linked-list.svg/500px-Singly-linked-list.svg.png)

#### Q 1.2.1
Consider a [singly linked list](https://en.wikipedia.org/wiki/Linked_list), $L$. Write a function `is_palindrome(L)` that detects if $L$ is a [palindrome](https://en.wikipedia.org/wiki/Palindrome), by returning a bool, `True` or `False`.


#### Q 1.2.2
What is the complexity of your implementation?

#### Q 1.2.3
Consider an alternative implementation to detect if L is a palindrome and write a new function, `is_palindrome2(L)`.

#### Q 1.2.4
What's the complexity of this implementation?


#### Q 1.2.5 
What are some examples of optimizations that could improve computational performance?


## 2. Prob + Stats

### Q 2.1: Finding $\pi$ in a random uniform?
<img src=https://www.epicurus.com/food/recipes/wp-content/uploads/2015/03/Pi-Day.jpg width="480">

Given a uniform random generator $[0,1)$ (e.g., use your language's standard libary to generate random value), write a a function `compute_pi` to compute [$\pi$](https://en.wikipedia.org/wiki/Pi).

In [26]:
import numpy as np

def compute_pi(n=100000):
    inside = 0
    for i in range(n):
        # generate coordinates uniformly in [0, 1)
        x = np.random.uniform()
        y = np.random.uniform()
        
        # count number of points that lie inside a quarter of a circle
        if np.sqrt(x**2 + y**2) <= 1:
            inside += 1

    pi = 4*inside/n
    return pi

compute_pi()

3.14876

### Q 2.2: Making a 6-side die roll a 7?

Using a single 6-side die, how can you generate a random number between 1 - 7?

### Q 2.3: Is normality uniform?

<img src=https://rednaxela1618.files.wordpress.com/2014/06/uniformnormal.png width="480">


Given draws from a normal distribution with known parameters, how can you simulate draws from a uniform distribution?

### Q 2.4: Should you pay or should you go?

![coin flip](https://lh5.ggpht.com/iwD6MnHeHVAXNBgrO7r4N9MQxxYi6wT9vb0Mqu905zTnNlBciONAA98BqafyjzC06Q=w300)

Let’s say we play a game where I keep flipping a coin until I get heads. If the first time I get heads is on the nth coin, then I pay you $2^{(n-1)}$ US dollars. How much would you pay me to play this game? Explain.

### Q 2.5: Uber vs. Lyft

![uber vs lyft](http://usiaffinity.typepad.com/.a/6a01347fc1cb08970c01bb0876bcbe970d-pi)

You request 2 UberX’s and 3 Lyfts. If the time that each takes to reach you is IID, what is the probability that all the Lyfts arrive first? What is the probability that all the UberX’s arrive first?

### Q 2.6: Pick your prize
<img src=https://miro.medium.com/max/1100/1*m5b3O9sE68UCXjLw5oxy2g.png width="480">

A prize is placed at random behind one of three doors and you are asked to pick a door. To be concrete, say you always pick door 1. Now the game host chooses one of door 2 or 3, opens it and shows you that it is empty. They then give you the option to keep your picked door or switch to the unopened door. Should you stay or switch if you want to maximize your probability of winning the prize?

## 3 Conceptual ML

### Q 3.1 Why study gradient boosting or neural networks?

Consider a regression setting where $X \in \mathbb{R}^p$ and $Y \in \mathbb{R}$. The goal is to come up with a function $f(X): \mathbb{R}^p \rightarrow \mathbb{R}$ that minimizes the squared-error loss $(Y - f(X))^2$. Since X, Y are random variables, we seek to minimize the expectation of the squared error loss as follows
\begin{equation}
EPE(f) = \mathbb{E}\left[(Y-f(X)^2\right]
\end{equation}
where EPE stands for expected prediction error. One can show that minimizing the expected prediction error leads to the following _regression function_
\begin{equation}
f(x) = \mathbb{E}\left[Y|X=x\right]
\end{equation}

The goal of any method is to approximate the regression function above, which we denote as $\hat{f}(x)$. For example, linear regression explicitly assumes that the regression function is approximately linear in its arguments, i.e. $\hat{f}(x) = x^T\beta$ while a neural network provides a nonlinear approximation of the regression function. 

The simplest of all these methods is [k-nearest neighbors](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm). Given $x$ and some neighbourhood of $k$ points $N_k(x)$, $\hat{f}(x)$ is simply the average of all $y_i|x_i \in N_k(x)$.  Let $N$ denote the training sample size. Under mild regularity conditions on the joint probability distribution $Pr(X, Y)$, one can show that as $N \rightarrow \infty$, $k \rightarrow \infty$ such that $k/N \rightarrow 0$, then $\hat{f}(x) \rightarrow f(x)$ where $\rightarrow$ means approaches or goes to. In other words, the k-nearest neighbors algorithm converges to the ideal solution as both the training sample size and number of neighbors increase to infinity.

Now given this _universal approximator_, why look any further and research other methods? Please share your thoughts.


### Q 3.2 Model Selection and Assesment

Consider a multiclass classification problem with a large number of features $p >> N$, for e.g $p=10000, N=100$ The task is threefold
1. Find a "good" subset of features that show strong _univariate_ correlation with class labels
2. Using the "good" subset, build a multi class classifier
3. Estimate the generalization error of the final model

Given this dataset, outline your approach and please be sure to cover the following
- Data splitting
- Model Selection: either estimating the performance of different classifiers or the same classifier with different hyperparameters
- Model Assessment: having chosen a classifier, estimating the generalization error

Assume all features are numerical, the dataset contains no NULLS, outliers, etc. and doesn't require any preprocessing.



### My Response

1. Find a "good" subset of features that show strong _univariate_ correlation with class labels

For each feature, we can conduct a One Way ANOVA test to determine whether there are any statistically significant differences between the means of the 100 classes. Features that have statistically significant differences between the means of the 100 classes are then used for modeling. If the means of the 100 classes are almost the same, it could be a good indication that this feature does not provide much information about the target variable, and thus should be discarded.

We can then perform forward/backward selection, or look at the feauture importances in tree-based models to further reduce the number of features and only include relevant features for training.

If the remaining subset of features after performing this test is still large, we can consider using PCA to reduce the dimensions of the features to get an even smaller subset of the features.

2. Using the "good" subset, build a multi class classifier

First, split the dataset into train, validation, and test dataset with 70:10:20 ratio. Make sure that the distribution of classes is the same across train, validation, and test dataset. If the original dataset is small, we can consider using K-fold cross validation, where we combine both train and validation dataset, split them into K folds, train a model using K-1 folds and validate the model using the Kth fold, and repeat this process untill every K-fold has been served as the test dataset. We then take the average of the error and this will serve as the performance metric for this model.

For model selection, we can first start with comparing the generalization errors calculated from validation dataset of different classifiers with multi-class loss functions such as multinomial logistic regression, support vector machine, decision trees, K-nearest neighbors, and naive bayes. We can also consider a simple neural network. We train each model with the same train and test dataset and choose the one with the least generalization error. We then perform hyperparameter tunining (with grid search, random search, or bayes optimization) on that model to further lower the generalization error. Ideally, the generalization errors evaluated on train and validation dataset should be low and similar to each other. If the generalization error evaluated on train dataset is high, it means that the model is underfitting and we need to tune our model, train longer, or choose other types of models. If the error on validation dataset is significantly higher than the error on train dataset, it means that the model is overfitting and we should try different methods (more training data, early stopping, regularization, etc.) to prevent overfitting.

3. Estimate the generalization error of the final model

For a multi-class classification problem, we can use the macro-averaged F1-score, which is a simple arithmetic mean of the per-class F1-scores, as an estimate to the generalization error of the final model. If the macro-averaged F1-score
evaluated on the test dataset is similar to the score evaluated on the validation dataset, it means that the model is able to generalize well to unseen data. Depending on the application of this multi-class classification, we can consider different loss functions (such as focal loss) and a weighted-average F1-score for evaluation, so that we pay more attention to a specific class.

Above is a general framework on how I would approach a multi-class classification problem. With more details and domain knowledge about the problem, we can definitely skip some of the steps mentioned (often to reduce computational costs and to come up with a MVP quickly) and choose to pay attention to other aspects of this framework. 